Arnab/swiper

arnab-roy opened the pull request:
This PR adds a "super_swiper" algorithm for weight reduction in the `fastcrypto-tbls` crate, adds supporting infrastructure for weight reduction experiments, and enhances the codebase with tools and data for working with real validator sets.

The super_swiper code is adapted from: https://github.com/tolikzinovyev/weight-reduction

The implementation of the algorithms presented in the paper:
Weight reduction in distributed protocols: new algorithms and analysis, by Anatoliy Zinovyev
(https://eprint.iacr.org/2025/1076).

arnab-roy committed:
Add super_swiper weight reduction and tests

Adds the optional 'super-swiper' feature to fastcrypto-tbls, adding dependencies on a local solver and num-rational. Implements Nodes::new_super_swiper_reduced for advanced weight reduction using the super_swiper algorithm, and provides a dedicated test for this functionality with realistic validator weights.

arnab-roy committed:
Add weight-reduction solver repo

Adds a copy of the 'weight-reduction' repo under fastcrypto-tbls, including solver algorithms, utilities, tests, and datasets. Includes Cargo.toml, license, documentation, and data files for reproducibility and usage as described in the referenced research paper.

arnab-roy committed:
Add Sui validator stake/ticket scripts and test integration

Added scripts to fetch all Sui validators and generate CSVs of stake weights and ticket assignments under src/weight-reduction/scripts. Integrated super_swiper_test.rs as a conditional test module and updated its path and imports. Added a new data file (sui_stakes_and_tickets.csv) for validator stake/ticket analysis.

arnab-roy committed:
Add Sui validator stake data files

Added sui_real_all.dat and sui_real_all_details.txt containing Sui validator stake amounts and corresponding validator names for use in weight reduction analysis.

arnab-roy committed:
Refactor super swiper reduced node creation API

Removed unused parameters 't' and 'total_weight_lower_bound' from Nodes::new_super_swiper_reduced and updated threshold calculation to use beta ratio directly. Updated tests to match new API and added a new test for additional weight scenarios.

arnab-roy committed:
Remove unused weight-reduction solvers and tests

Deleted several solver implementations and their associated tests from the weight-reduction module, including bruteforce_sorted, faster_swiper, gcd_wr, and swiper. Updated mod.rs to only include super_swiper. This streamlines the codebase by removing unused or deprecated algorithms and their test files.

arnab-roy committed:
Remove weight-reduction data and test files

arnab-roy committed:
Integrating Ratio types

arnab-roy committed:
Removing swiper folder

arnab-roy committed:
More cleanup

arnab-roy committed:
More cleanups

arnab-roy committed:
More cleanups

arnab-roy committed:
More cleanups

arnab-roy committed:
fixed nightly build req

arnab-roy committed:
Update super_swiper_test.rs

arnab-roy committed:
Remove super-swiper feature gating and update dependencies

The 'super-swiper' feature flag and related optional dependencies were removed, making the super_swiper functionality always available.

arnab-roy committed:
Clean up deps

arnab-roy committed:
Add weight reduction validation and enhance super_swiper

Introduces the weight_reduction_checks module for validating weight reduction properties, and updates Nodes::new_super_swiper_reduced to iteratively adjust beta to meet a total weight upper bound and validate reductions. Test cases are expanded to cover new validation logic, CSV output, and upper bound scenarios.

arnab-roy committed:
Refactor super swiper reduction to use slack-based validation

Updated Nodes::new_super_swiper_reduced to use a slack-based approach for weight reduction validation, replacing the previous total weight upper bound check. The function now iterates beta to satisfy a configurable slack constraint, and returns additional beta parameters. Tests and weight_reduction_checks were updated to support and verify the new slack-based logic.

arnab-roy committed:
Cleanup

arnab-roy committed:
GH tests

jonas-lj committed:
lock

jonas-lj committed:
Refactor

jonas-lj committed:
revert

arnab-roy committed:
Refactor for GH tests

arnab-roy committed:
Edits for fmt check

arnab-roy committed:
Edits for ubuntu GH actions

arnab-roy committed:
Edits for Ubuntu GH actions

arnab-roy committed:
Refactor super_swiper_reduced to use threshold t

Changed the Nodes::new_super_swiper_reduced function to accept an absolute threshold 't' instead of an alpha ratio, computing alpha internally. Updated related test to pass 't' and clarified output for threshold and alpha calculation.

arnab-roy committed:
Refactor weight reduction to use delta constraint

Replaces slack-based validation with delta-based checks in super swiper weight reduction logic. Aligns interface with new_reduced.

arnab-roy committed:
Add test for new_reduced with delta constraint

Introduces a test for the new_reduced function using an 8% delta constraint on the original total weight. The test verifies weight reduction, threshold adjustment, node ID preservation, and writes results to a CSV file for further analysis.

arnab-roy committed:
Add Sui validator voting power data and test integration

Added a script to fetch all Sui validators' voting power and save it to a new data file. Integrated the new voting power data into super_swiper_test.rs, updating tests to use the new dataset and output results to new CSV files.

arnab-roy committed:
Add multi-epoch Sui voting power data and tests

Introduces voting power data files for epochs 100, 200, 400, 800, and renames the previous file to epoch_974. Updates test helpers and adds a comprehensive test to compare weight reduction algorithms across all epochs. The fetch script is enhanced to support epoch selection via command line and outputs files with epoch-specific names.

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+    }
+
+    fn get(&self, index: usize) -> u64 {
+        match self.tickets.get(index) {
```
This is the same as `unwrap_or`

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+    // Returns the maximum achievable adversarial number of tickets.
+    #[cfg(test)]
+    fn adversarial_tickets(&self) -> u64 {
+        for (t, &w) in self.dp.iter().enumerate().rev() {
```
You can use `find_position` here.

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+    indices_tail.sort_unstable_by(|a, b| b.cmp(a));
+    indices_tail.dedup();
+
+    let mut indices_head: Vec<_> = (0..tickets_len).collect();
```
Using swap_remove here may modify the order of the remaining elements, so is this safe? Also, isn't there a better way to do this instead of creating a long vector and then remove elements?

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+    let mut g = generate_deltas(weights, alpha);
+
+    let mut batch_size: usize = 1;
+    loop {
```
Can you write a note about why this doesn't run forever?

arnab-roy replied:
Added explanation.

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+                    if tt >= adv_tickets_target {
+                        return None;
+                    }
+                    while self.dp.len() <= tt {
```
Use `resize` here

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+        }
+
+        let mut dp = Vec::with_capacity(adv_tickets_target);
+        for &x in self.dp.iter().take(adv_tickets_target) {
```
You can just collect the iterator as `dp` here

arnab-roy commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+
+        for i in (1..self.dp.len()).rev() {
+            if self.dp[i] != 0 {
+                let ww = self.dp[i] + w;
```
Renamed to be clearer.

arnab-roy commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+        0
+    }
+    fn adv_tickets_target(&self) -> u64 {
+        self.dp.capacity() as u64
```
Added as an attribute.

arnab-roy commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+    Some(dp)
+}
+
+fn update_tickets(deltas: &[usize], tickets: &mut Tickets) {
```
Refactored update_tickets, but not sure about batch_recursive as it's coordinating many structures.

arnab-roy committed:
Resolving comments from Jonas

arnab-roy committed:
fmt and clippy fixes

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,539 @@
+    i: usize,
+}
+
+impl Ord for QueueElement {
```
In that case, you can just switch the order of the fields in the struct definition.

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,552 @@
+            return tickets.into_vec();
+        }
+
+        batch_size *= 2;
```
Is there an upper bound on the number of iterations which guarantees that this doesn't overflow?

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,552 @@
+/// the new ticket assignment after applying all deltas.
+/// `dp_head` must have all indices applied except those in `deltas` with data
+/// in `tickets`.
+fn process_batch_recursive(
```
Rust has a max call depth, so we need some upper bound on the number of recursions to make sure this will not panic - do we have that?

jonas-lj commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,552 @@
+            .map(|t| t as u64)
+            .unwrap_or(0)
+    }
+    fn adv_tickets_target(&self) -> u64 {
```
It doesn't seem like this function is used anywhere?

arnab-roy replied:
It's used in line 354

arnab-roy commented on fastcrypto-tbls/src/weight_reduction/mod.rs:
```diff
@@ -0,0 +1,552 @@
+
+    fn update(&mut self, index: usize) {
+        if index >= self.tickets.len() {
+            self.tickets.resize(index + 1, 0);
```
Removing the if check and calling resize(index + 1, 0) unconditionally would truncate the vector when index < self.tickets.len(), causing data loss and "N/A" results.

arnab-roy committed:
Ensuring and clarifying loop bounds

jonas-lj committed:
Misc clean up

jonas-lj committed:
Align names with paper

jonas-lj committed:
Simplify'ish + refactor

jonas-lj committed:
Simplify a bit in new_reduced function

jonas-lj committed:
more simplifications

arnab-roy committed:
New precision loss calculation and delta check

Refactored the super_swiper weight reduction logic to use a new precision loss calculation (compute_precision_loss) instead of the previous delta check, and updated the iteration logic for beta selection. Added new test utilities and CSV generation for per-epoch validator weight reductions, and improved comparison and reporting in tests. Introduced compute_precision_loss in weight_reduction_checks.rs for more accurate measurement of reduction precision loss.

arnab-roy committed:
Document and refactor super_swiper weight reduction logic

Added detailed DKG safety and liveness constraints to the super_swiper algorithm in nodes.rs, referencing a new PROOF.md with formal definitions and proofs. Refactored the API and comments for clarity, updated test types and output in super_swiper_test.rs, and improved documentation and comments in weight_reduction_checks.rs to match the formal proof. No functional changes to the core reduction logic.

arnab-roy commented on fastcrypto-tbls/src/weight_reduction/weight_reduction_checks.rs:
```diff
@@ -0,0 +1,194 @@
+/// 1. Take reduced weights to reach 2t' - 1
+/// 2. Let w1 = sum of old weights corresponding to the subset
+/// 3. Then delta = w1 - (2t - 1), where t is the input threshold
+pub fn get_delta(
```
Kept it there for future testing, but we can remove once confident in the precision loss based parameter tuning.

arnab-roy committed:
Updated t', f' constraints

PROOF.md expanded with formal upper/lower bound proofs and a derived combined inequality for admissible f', t', and W' to match the implementation changes. The algorithm uses the precision loss (δ) and effective divisor (d) to derive tight lower/upper bounds for f' (from PROOF.md), selects a valid integer f' and verifies all constraints while iterating β. Tests updated to expect and display the extra f' and β info.

arnab-roy committed:
Update beta and f' calculations

Replace the previous iterative search for beta with a direct formula beta = alpha + (allowed_delta - 1)/W and simplify the super_swiper weight reduction logic accordingly.

arnab-roy committed:
Updated the analysis and modified node.rs

arnab-roy committed:
Simplify super_swiper reduction and update tests

Refactor Nodes::new_super_swiper_reduced to remove the f parameter and f' calculation, change t' formula to (t + δ)/d, and return (Nodes, t') instead of (Nodes, t', f'). Use original_total_weight consistently for ratio calculations and tighten the beta loop condition (beta must be > α). Update call sites and construction accordingly.

Revamp the super_swiper tests: replace the single ad-hoc test with a comprehensive epoch-comparison suite that supports both new_super_swiper_reduced and new_reduced, introduces helpers for epoch charting, f/f' derivation, and precision-loss reporting, and adds multiple test cases and printing improvements to validate the new behavior and signatures.

arnab-roy committed:
Refine super_swiper weight reduction search

Replace the single-β loop with a double loop over α (0.10..0.90) and β (α+0.01..α+0.20, β<1). For each (α,β) the code calls solve, computes reduced total weight and precision loss, and selects the best feasible reduction (2·δ ≤ allowed_delta and W' ≥ total_weight_lower_bound). Tie-breaking prefers smaller W', then smaller δ. The old α computation and early-return-on-below-bound behavior were removed so invalid β values are skipped instead of failing. Also shorten test labels in super_swiper tests to simpler names.

arnab-roy committed:
Update weight_reduction PROOF: bounds & params

Align analysis with concrete examples

arnab-roy committed:
Update PROOF.md

Add guarantees.

arnab-roy committed:
Merge branch 'main' into arnab/swiper

arnab-roy committed:
Refactor super-swiper and weight reduction

Clean up and clarify super-swiper and weight-reduction internals: introduce a SuperSwiperBest type alias and standardize tuple field order to avoid mixing delta and weights; simplify selection logic and return handling. Move neg_mod into the Nodes impl as an associated function and update its call sites. Include accumulated_weights and nodes_with_nonzero_weight in Nodes construction. Use div_ceil for left-branch sizing in batch processing and replace manual Option matches with the ? operator in weight reduction checks. Also apply minor formatting and test cleanup for readability.

arnab-roy committed:
Add new_reduced_v2 weight-reduction and tests

Introduce Nodes::new_reduced_v2 implementing a floor-division based weight reduction algorithm that scans coarse and fine divisor candidates, computes exact precision loss via compute_precision_loss, and returns reduced Nodes with computed t' and f' (unilateral formulas). Add a ReductionBest type alias and ensure total-weight/liveness/δ constraints are enforced. Also add run_all_epochs_v2_chart and two unit tests (test_all_epochs_new_reduced_v2, test_all_epochs_new_reduced_v2_t52_l80_delta8) to exercise the new reduction across multiple epochs and report comparison metrics.

arnab-roy committed:
Add W vs W' comparison chart test

Add a helper (run_all_epochs_w_and_wprime_by_algorithm_chart) to compute and print W and W' for epochs [100,200,400,800,974], comparing three reduction algorithms (new_reduced, new_super_swiper_reduced, new_reduced_v2). The function scales validator weights, runs each reduction, prints a formatted table, and asserts each algorithm produced a result. Also add test_all_epochs_w_and_wprime_by_algorithm_chart to run this check with baseline epoch parameters.

benr-ml commented:
@arnab-roy let's remove everything related to super_swiper since we are not planning to use it

arnab-roy committed:
SuperSwiper: bilateral analysis and tests

Change Nodes::new_super_swiper_reduced to implement the bilateral weight-reduction analysis: accept original-space f and return (reduced Nodes, t', f'). Replace the previous feasibility check (2·δ ≤ allowed_delta) with the bilateral criterion 3·δ + d ≤ allowed_delta, compute closed-form reduced thresholds t' = ⌈(t+δ)/d⌉ and f' = ⌊(f+δ)/d⌋, and update docs to describe the two-stage search/closed-form procedure. Adjust internal weight/delta computations and refactor some local variables. Extend tests heavily: add helpers for reducers that take/return f, derive f from chart params, add run_all_epochs_with_f_chart and side-by-side comparison tests between new_super_swiper_reduced and new_reduced_with_f, and adapt existing tests to the new return shape and semantics.
